Change current curl fork/exec with libdavix plugin for read operations

juztas opened the pull request:
- disables http/web access for curl and use Davix library.
  - Implemented read, vector read and + offset,
  - Authentication check 
  - Allow to control Davix debug mode.

http/https/web was tested by turning on http[s] access for an xrootd server. More tests are upcoming to identify the gain/loss.
This requires davix 0.6.3 (https://github.com/cms-sw/cmsdist/pull/2355)

dtnrm committed:
Disable web/http protocol for curl storage plugin

dtnrm committed:
New DavixAdaptor plugin skeleton
Support http/web protocols and implemented basic functions: open, read, position

dtnrm committed:
Davix close/abort function

dtnrm committed:
Implement Davix Vector read operation

dtnrm committed:
Implement Davix Vector read operation with offset option

dtnrm committed:
Authentication priority check:
  1. X509_USER_PROXY
  2. /tmp/x509up_u<(geteuid)>
  3. X509_USER_{CERT|KEY}
If none of these is present, it will try to make connection without proxy/cert. X509_CERT_DIR can also be overwritten with environment variable.

dtnrm committed:
Davix_Debug is controlled with flags [0-4] through env variable:
  0. Default davix log level;
  1. DAVIX_LOG_WARNING;
  2. DAVIX_LOG_VERBOSE;
  3. DAVIX_LOG_DEBUG;
  4. DAVIX_LOG_ALL;
If provided value is not integer or it is not able to parse to int, it will use 0. Any other number from 4toN will be considered as 4. DAVIX_LOG_ALL which is most verbose;

dtnrm committed:
Improve logging for DavixAdaptor plugin in all functions
It checks DavixErr if it was set or not also if FD was opened sucessfully and if return value is greater than 0. 0 - means it is end of file. If any of these are found it will raise an issue.

dtnrm committed:
Close Davix_FD before opening new one

cmsbuild commented:
A new Pull Request was created by  @juztas (Justas Balčas) for CMSSW_8_1_X.

It involves the following packages:

Utilities/DavixAdaptor
Utilities/StorageFactory


The following packages do not have a category, yet:

Utilities/DavixAdaptor

@cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

juztas commented:
Reference for more information: https://indico.cern.ch/event/541112/contributions/2212365/attachments/1295742/1931898/Http_Data_Federations_in_CMS_Project20160621JBalcas.pdf
@bbockelm ;)

bbockelm commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,346 @@
+
+IOSize DavixFile::write(const void *from, IOSize n) {
+  cms::Exception ex("FileWriteError");
+  ex << "DavixFile::write(name='" << m_name << "') not implemented";
```
Can you do `ex.addContext("Calling DavixFile::write()");` here?

bbockelm commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,346 @@
+       << s;
+    ex.addContext("Calling DavixFile::readv()");
+    throw ex;
+  } else if (s == 0)
```
Surround with braces `{}` as you do with the rest of the conditions.

bbockelm commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,346 @@
+  } else if (s == 0)
+    return 0; // end of file
+
+  return total;
```
What's the meaning of the return value, `s`?  Is it the number of bytes read?  Do we need a check that `s <= total`?

bbockelm commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,346 @@
+  }
+
+  ssize_t s = davixPosix->preadVec(m_fd, input_vector, output_vector, buffers, &davixErr);
+  if (davixErr) {
```
What's the memory management of `davixErr`?  Is this a memory leak?

Use `std::unique_ptr` if we need to own the lifetime of this object.

Dr15Jones replied:
I checked the Davix code and we are responsible for deleting the `DavixError` if it is set.

bbockelm commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,346 @@
+    throw ex;
+  }
+
+  if (davixPosix && m_fd)
```
Calling `open()` on an open file should cause an exception, not silently ignored.

bbockelm commented on Utilities/DavixAdaptor/interface/DavixFile.h:
```diff
@@ -0,0 +1,49 @@
+  Davix_fd *m_fd;
+  Davix::Context *davixContext;
+  Davix::DavPosix *davixPosix;
+  Davix::RequestParams *davixReqParams;
```
Make owned pointers `std::unique_ptr` as applicable in order to avoid memory ownership issues and leaks.

dtnrm committed:
Add missing BuildFile for Davix plugin

dtnrm committed:
DavixAdaptor unique_ptrs, global var, null fixes
Do not use global variable for Davix Context
Use unique_ptr in places where it is possible;
Define m_name before throwing an error, otherwise it is null
Check variables without NULL.

dtnrm committed:
Update Davix exceptions, braces, compare return bytes
Throw exception if open is called on an open file
Compare bytes requested vs read and throw exceptions
Surround with braces {} as done in the rest of the code

cmsbuild commented:
Pull request #15130 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

bbockelm commented:
This is beginning to look good to me.  Can you provide an overview of how you've tested this?

Thanks.

smuzaffar commented:
@juztas , can you please add few unit tests ?

dtnrm committed:
DavixError has to be NULL and we can't check vector read return
1)During initialization if we want to check and get DavixError it has to be NULL;
2) Currently Davix limits number of requests sent to the server in the interest of improving performance. This means that we can`t rely on value returned by preadVec if buffers > 1. It would work only in few cases, whenever ranges are not coalescing. If two ranges are overlapping by a single byte, it will contain one byte less in return.

cmsbuild commented:
Pull request #15130 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

dtnrm committed:
Add unit tests for Davix plugin
This pull request makes sure that old test which are on StorageFactory/test will work as it should be.
Also new tests are checking if file is accessible, it can do read, vector read with offset and also check if filenotfound or wrong url will raise proper errors.

cmsbuild commented:
Pull request #15130 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

juztas commented:
@smuzaffar added. Currently afaik only Caltech has exported http access through xrootd. Files which are read are HC dataset which should be always on all sites. I also made sure that old http tests from StorageFactory will work as it should.  
@bbockelm I will write down a twiki with all my done tests with diff cacheHint's and with info about debug mode and certificates priority and flags.

smuzaffar commented:
@juztas , there are few unit tests in https://github.com/cms-sw/cmssw/tree/CMSSW_8_1_X/Utilities/StorageFactory/test which are using http storagefactory. As http/web plugins are provided by David package , so may be those tests should be moved here too?

dtnrm committed:
Move http tests from Utilities/StorageFactory to Utilities/DavixAdaptor

cmsbuild commented:
Pull request #15130 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

juztas commented:
Sure, moved

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 6b09c170e4a90ee9a147a5143293bbb549e3d79b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15130/14159/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15130/14159/summary.html

davidlt commented:
ping^1

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+  if (buffers == 0)
+    return 0;
+
+  DavixError *davixErr = NULL;
```
please use `nullptr` since `NULL` is a C construct not C++.

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+  DavixError *davixErr = NULL;
+
+  DavIOVecInput input_vector[buffers];
+  DavIOVecOuput output_vector[buffers];
```
@davidlt Since this is not standard C++ is this going to be a problem for clang?

davidlt replied:
@Dr15Jones I think, @juztas already was using Clang during the development to iron out possible issues and it did compile. Clang does support VLA (.. or to some undocumented scope). I think, Davix examples also contains VLAs.

We could replace them with allocation on heap instead and VLAs are gone.

bbockelm replied:
Seems that it's just safer to switch to standard C++?

davidlt replied:
Yes. It's preferred.

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+
+  if (davixPosix && m_fd) {
+    edm::Exception ex(edm::errors::FileOpenError);
+    ex << "Davix::open(name='" << m_name->c_str() << "') failed on already open file";
```
replace `m_name->c_str()` with `*m_name` since there is no need to force the existence of the C string.

Dr15Jones replied:
It doesn't look like `m_name` is even set yet since that happens at line 186.

Dr15Jones replied:
I suggest making m_name just a `std::string` since it will make memory management easier. I'd also set `m_name = name` at line 152 right after you've tested that `name` is valid.

Dr15Jones commented on Utilities/DavixAdaptor/interface/DavixFile.h:
```diff
@@ -0,0 +1,50 @@
+  // Because Davix_fd is not available in C++ header files and
+  // sizeof cannot with incomplete types
+  Davix_fd *m_fd;
+  std::unique_ptr<Davix::DavPosix> davixPosix;
```
please change to `m_davixPosix` to match the other variables and standard CMS naming conventions.

Dr15Jones commented on Utilities/DavixAdaptor/interface/DavixFile.h:
```diff
@@ -0,0 +1,50 @@
+  // sizeof cannot with incomplete types
+  Davix_fd *m_fd;
+  std::unique_ptr<Davix::DavPosix> davixPosix;
+  std::unique_ptr<Davix::RequestParams> davixReqParams;
```
please change to `m_davixReqParams`.

Dr15Jones commented on Utilities/DavixAdaptor/interface/DavixFile.h:
```diff
@@ -0,0 +1,50 @@
+  Davix_fd *m_fd;
+  std::unique_ptr<Davix::DavPosix> davixPosix;
+  std::unique_ptr<Davix::RequestParams> davixReqParams;
+  std::unique_ptr<std::string> m_name;
```
As mentioned in another comment, please change to `std::string m_name`

Dr15Jones commented on Utilities/DavixAdaptor/plugins/DavixStorageMaker.cc:
```diff
@@ -0,0 +1,50 @@
+    davixPosix->stat64(NULL, newurl, &info, &err);
+    if (err) {
+      std::unique_ptr<Davix::DavixError> davixErrManaged(err);
+      std::cout << "Check failed with error " << err->getErrMsg().c_str() << " and error code"
```
`std::cout` is not allowed in this context. Please use MessageLogger.

Dr15Jones replied:
Actually, better to just remove this debugging all together since the same information is encoded in the exception.

Dr15Jones commented on Utilities/DavixAdaptor/plugins/DavixStorageMaker.cc:
```diff
@@ -0,0 +1,50 @@
+      ex.addContext("Calling DavixFile::check()");
+      throw ex;
+    }
+    std::cout << "File is " << info.size << " bytes large" << std::endl;
```
Please remove debugging statements.

Dr15Jones commented on Utilities/DavixAdaptor/test/davixcheck.cpp:
```diff
@@ -0,0 +1,22 @@
+                                             &size);
+
+  std::cout << "exists = " << exists << ", size = " << size << "\n";
+  std::cout << "stats:\n" << StorageAccount::summaryText() << std::endl;
```
Please remove these `std::cout` statements or move them to `MessageLogger`.

Dr15Jones replied:
Ah, I didn't noticed this was in a test. `cout` in tests are fine.

Dr15Jones commented on Utilities/DavixAdaptor/interface/DavixFile.h:
```diff
@@ -0,0 +1,50 @@
+#include "Utilities/StorageFactory/interface/Storage.h"
+#include <davix.hpp>
+
+class DavixFile : public Storage {
```
Please add `override`  to each of the virtual functions you are implementing for the base class `Storage`. This will help guarantee there is no typos.

Dr15Jones commented on Utilities/DavixAdaptor/plugins/DavixStorageMaker.cc:
```diff
@@ -0,0 +1,50 @@
+                     IOOffset *size = 0) const override {
+    std::string newurl((proto == "web" ? "http" : proto) + ":" + path);
+    Davix::DavixError *err = NULL;
+    std::unique_ptr<Davix::DavPosix> davixPosix =
```
There doesn't appear to be a reason to use a `std::unique_ptr<Davix::DavPosix>` here. Instead just create one on the stack
```C++
Davix::DavPosix davixPosix;
```

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+void DavixFile::close(void) {
+  if (davixPosix && m_fd) {
+    DavixError *err;
+    davixPosix->close(m_fd, &err);
```
What happens if `err` is non-null?

Dr15Jones replied:
If `err` is non-null this would be a memory leak.

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+
+void DavixFile::close(void) {
+  if (davixPosix && m_fd) {
+    DavixError *err;
```
`DavixError* err = nullptr;` since it is unclear what the API will do if you pass a non-null.

Dr15Jones replied:
I looked at the implementation of Davix and `DavixError` must always be `nullptr` when calling its API.

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+
+void DavixFile::abort(void) {
+  if (davixPosix && m_fd) {
+    DavixError *err;
```
Again, set to `nullptr` and what happens if there is an error?

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+
+void DavixFile::configureDavixLogLevel() {
+  long logLevel = 0;
+  char *logptr;
```
`nullptr`

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+void DavixFile::configureDavixLogLevel() {
+  long logLevel = 0;
+  char *logptr;
+  char *davixDebug = getenv("Davix_Debug");
```
This can be a `char const * const`

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+    edm::LogWarning("DavixFile") << "Was not able to find proxy in $X509_USER_PROXY, "
+                                 << "X509_USER_{CERT|KEY} or default proxy creation location. "
+                                 << "Will try without authentication";
+    return -1;
```
Does `davixErr` need to be set to something? At very least `nullptr`?

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+    openflags |= O_RDONLY;
+
+  DavixError *davixErr = NULL;
+  davixReqParams = std::make_unique<RequestParams>();
```
It doesn't look like this needs to be a `std::unique_ptr`. Instead just use an object on the stack
```C++
RequestParams davixReqParams;
```

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+  DavixError *davixErr = NULL;
+
+  DavIOVecInput input_vector[buffers];
+  DavIOVecOuput output_vector[buffers];
```
Again with the non standard C++.

Dr15Jones replied:
Again with the non-standard VLA.

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+  if (buffers == 0)
+    return 0;
+
+  DavixError *davixErr = NULL;
```
`nullptr`

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+  // Only check if returned val <= 0 and make proper actions.
+  if (s < 0) {
+    edm::Exception ex(edm::errors::FileReadError);
+    ex << "Davix::readv(name='" << m_name->c_str() << "', n=" << buffers
```
No need for calling `m_name->c_str()` especially when `m_name` is no longer a `std::unique_ptr`.

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+}
+
+IOSize DavixFile::read(void *into, IOSize n) {
+  DavixError *davixErr = NULL;
```
`nullptr`

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+}
+
+IOOffset DavixFile::position(IOOffset offset, Relative whence /* = SET */) {
+  DavixError *davixErr = NULL;
```
`nullptr`

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+void DavixFile::abort(void) {
+  if (davixPosix && m_fd) {
+    DavixError *err;
+    davixPosix->close(m_fd, &err);
```
Need to clean up memory if `err` is not `nullptr`.

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+  }
+  switch (logLevel) {
+  case 0:
+    davix_set_log_level(0);
```
I checked and it turns out this is not technically thread safe because the underlying global variable being set is not `atomic`.
https://github.com/cern-it-sdc-id/davix/blob/25976712a9c88277894324f6280f23e5b5955958/src/utils/davix_logger.cpp#L33

In principal I'm not worried about this under intel like machines but could be a problem under other architectures.

bbockelm replied:
I don't think this is an appropriate place to be setting the log level: it's much better in the plugin configuration, no?

At the plugin loading layer, the thread-safety concerns are lessened as we won't be loading multiple copies of `libdavix` in CMSSW (for now!).

We still ought to report the issue to upstream.  They've been friendly and happy to fix issues as reported.

davidlt replied:
PowerPC and ARMv8 with their weak memory model.

juztas replied:
Any comments on this? Leave for now as it is and report the issue to davix devs?

Dr15Jones replied:
It should be reported to Davix developers but we should protect this routine using a `std::call_once`. See
http://en.cppreference.com/w/cpp/thread/call_once

Dr15Jones commented on Utilities/DavixAdaptor/src/DavixFile.cc:
```diff
@@ -0,0 +1,363 @@
+  for (IOSize i = 0; i < buffers; ++i) {
+    input_vector[i].diov_offset = into[i].offset();
+    input_vector[i].diov_size = into[i].size();
+    input_vector[i].diov_buffer = (char *)into[i].data();
```
It isn't a good idea to use the old 'C' style casts. Would be better to do `static_cast<char*>(into[i].data())`. If that doesn't work than use a `reinterpret_cast<char*>(...)`

Dr15Jones commented:
I've finished my comments on the code.

dtnrm committed:
Update Davix plugin (nullptr, use vector, style casts, override)
DavixError must always be nullptr;
Override to each virtual function for the base class;
Use CMS naming conventions and DavixRequestParams no need to be private. Only used during open.
Remove left debugging statements;
During abort/close check davixError and in case it exists, raise exception.
Use std::vector for DavIOVecInput & DavIOVecOuput
Do not use old C' style casts and do static_cast<>()...

cmsbuild commented:
Pull request #15130 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 29aa4081d341cbe13c9e43bda3b9b2d5d77be6a2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15130/14264/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15130/14264/summary.html

dtnrm committed:
Make it thread safe calling davix_set_log_level

cmsbuild commented:
Pull request #15130 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 9221af13b71d0546764312be20f4b010d580b9b2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15130/14280/summary.html

Dr15Jones commented:
@bbockelm any further comments? I think I've finished.

bbockelm commented:
I'm happy!

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @smuzaffar

davidlange6 commented:
+1

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15130/14280/summary.html

smuzaffar commented:
@juztas , there are couple of unit test failing 
```
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc530/CMSSW_8_1_X_2016-08-09-2300/unitTestLogs/Utilities/DavixAdaptor
```
Can you please fix these? Looks like URL 
```
http://transfer-8.ultralight.org:1094/store/mc/HC/GenericTTbar/GEN-SIM-RECO/CMSSW_7_0_4_START70_V7-v1/00000///10FB32C7-0BCD-E311-B035-02163E00E79D.root
```
is not active any more.

juztas commented:
@smuzaffar I think it is more stable to use opendata.cern.ch as it also affected update to newest davix: https://github.com/cms-sw/cmsdist/pull/2454
Pull request here:
https://github.com/cms-sw/cmssw/pull/15579

I am running some tests on this node (transfer-8), so it is better to stay away from it for now.
